pkg/compactor/deleted_series_cleaner.go

```diff
@@ -0,0 +1,462 @@
+			Help: "Total number of blocks that need to be rewritten to processes the series deletion request.",
+		}, []string{"user"}),
+		tenentBlocksProcessedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_compactor_blocks_processed_for_series_deletion_total",
+			Help: "Total number of blocks processed for series deletion.",
+		}, []string{"user"}),
+		tenantSamplesProcessedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_compactor_samples_processed_during_series_deletion",
+			Help: "Total number of samples processed during series deletion, includes both samples reamining and deleted.",
+		}, []string{"user"}),
+		tenantSamplesDeletedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_compactor_samples_deleted",
+			Help: "Total number of samples deleted during series deletion.",
+		}, []string{"user"}),
+	}
+	c.Service = services.NewTimerService(cfg.CleanupInterval, c.ticker, c.ticker, nil)
+
+	return c
+}
+
+func (c *DeletedSeriesCleaner) ticker(ctx context.Context) error {
+	level.Info(c.logger).Log("msg", "started series deletion cleaner")
+	c.runsStarted.Inc()
+
+	err := c.cleanDeletedSeries(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to run the series deletion cleaner", "err", err.Error())
+		c.runsFailed.Inc()
+	} else {
+		level.Info(c.logger).Log("msg", "successfully ran the series deletion cleaner")
+		c.runsCompleted.Inc()
+		c.runsLastSuccess.SetToCurrentTime()
+	}
+
+	return nil
+}
+
+func (c *DeletedSeriesCleaner) cleanDeletedSeries(ctx context.Context) error {
+	users, deleted, err := c.usersScanner.ScanUsers(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to discover users from bucket")
+	}
+
+	isActive := util.StringsMap(users)
+	isDeleted := util.StringsMap(deleted)
+	allUsers := append(users, deleted...)
+
+	// Delete per-tenant metrics for all tenants not belonging anymore to this shard.
+	// Such tenants have been moved to a different shard, so their updated metrics will
+	// be exported by the new shard.
+	for _, userID := range c.lastOwnedUsers {
+		if !isActive[userID] && !isDeleted[userID] {
+			c.tenantPendingBlocks.DeleteLabelValues(userID)
+			c.tenantPendingRequests.DeleteLabelValues(userID)
+			c.tenantSamplesProcessedTotal.DeleteLabelValues(userID)
+			c.tenantSamplesDeletedTotal.DeleteLabelValues(userID)
+			c.tenentBlocksProcessedTotal.DeleteLabelValues(userID)
+			c.tenentTombstonesProcessedTotal.DeleteLabelValues(userID)
+		}
+	}
+	c.lastOwnedUsers = allUsers
+
+	return concurrency.ForEachUser(ctx, users, c.cfg.CleanupConcurrency, func(ctx context.Context, userID string) error {
+		_, err = c.deleteSeriesForUser(ctx, userID)
+		return errors.Wrapf(err, "failed to run the series deletion cleaner for user: %s", userID)
+	})
+}
+
+func (c *DeletedSeriesCleaner) deleteSeriesForUser(ctx context.Context, userID string) ([]ulid.ULID, error) {
```

Comment:
nit: When a func has grown very large, it's usually good to refactor it into smaller functions to make it readable. You could do something like this (just my suggestion): 

- `deleteSeriesForUser()`
  - top-level function


- `getRequestsPendingDeletion()`
  - Contains logic to list all deletion-requests and identify which are pending.
  - I'd add this function in the other file similar to: `GetAllDeleteRequestsForUser()`

- `getDeletionRequestsForBlock()`
  - Put the logic of checking which deletion-requests apply for a block

- etc